File: 368.largest-divisible-subset.js
/*
 * @lc app=leetcode id=368 lang=javascript
 *
 * [368] Largest Divisible Subset
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[]}
 */
const largestDivisibleSubset = function (nums)
{
    nums.sort((a, b) => a - b);

    /**
     * 所有的子集
     * @type {number[][]}
     */
    const subsets = [];
    /**
     * 最长子集的长度
     */
    let maxLength = 0;
    /**
     * 指向最长子集的引用
     * @type {number[]}
     */
    let maxLengthSubset = [];

    for (const num of nums)
    {
        let hasDivisibleSubset = false;
        // 因为会改变数组，因此要保存长度
        const subsetLength = subsets.length;
        /**
         * 对于 `num` 可以添加到末尾的最长子集的长度
         */
        let subsetMaxLength = 0;
        /**
         * 对于 `num` 可以添加到末尾的最长子集
         */
        let subsetWithMaxLength = [];
        for (let i = 0; i < subsetLength; i++)
        {
            const subset = subsets[i];
            const lastNum = subset[subset.length - 1];
            if (num % lastNum === 0)
            {
                hasDivisibleSubset = true;
                if (subset.length > subsetMaxLength)
                {
                    subsetMaxLength = subset.length;
                    subsetWithMaxLength = subset;
                }
            }
        }

        if (hasDivisibleSubset)
        {
            // 复制一份，因为可能还有其他因数的数字拼接
            subsets.push([...subsetWithMaxLength]);
            subsetWithMaxLength.push(num);
            if (subsetWithMaxLength.length > maxLength)
            {
                maxLength = subsetWithMaxLength.length;
                maxLengthSubset = subsetWithMaxLength;
            }
        }
        else    // 没有可以添加到末尾的子集，是新子集
        {
            const newSubset = [num];
            subsets.push(newSubset);
            if (1 > maxLength)
            {
                maxLength = 1;
                maxLengthSubset = newSubset;
            }
        }
    }
    return maxLengthSubset;
};
// @lc code=end

File: src/2025/368.largest-divisible-subset.cpp
/*
 * @lc app=leetcode id=368 lang=cpp
 *
 * [368] Largest Divisible Subset
 */

#include <algorithm>
#include <vector>

// @lc code=start
class Solution {
 public:
  std::vector<int> largestDivisibleSubset(std::vector<int>& nums) {
    std::ranges::sort(nums);
    std::vector<DpInfo> dp(nums.size());
    for (int i = 0; i < nums.size(); i++) {
      dp[i].prev_index_ = i;
      dp[i].subset_length = 1;
    }

    int max_subset_length = 1;
    int max_subset_length_index = 0;

    for (int j = 0; j < nums.size(); j++) {
      for (int i = j - 1; i >= 0; i--) {
        if (nums[j] % nums[i] == 0) {
          if (dp[i].subset_length + 1 > dp[j].subset_length) {
            dp[j].prev_index_ = i;
            dp[j].subset_length = dp[i].subset_length + 1;

            if (dp[j].subset_length > max_subset_length) {
              max_subset_length = dp[j].subset_length;
              max_subset_length_index = j;
            }
          }
        }
      }
    }

    return GetSubSetFromDpInfo(max_subset_length_index, dp, nums);
  }

 private:
  struct DpInfo {
    int prev_index_ = -1;
    int subset_length = -1;
  };

  static std::vector<int> GetSubSetFromDpInfo(const int index,
                                              const std::vector<DpInfo>& dp,
                                              const std::vector<int>& nums) {
    int current_index = index;
    std::vector<int> subset;
    while (dp[current_index].prev_index_ != current_index) {
      subset.push_back(nums[current_index]);
      current_index = dp[current_index].prev_index_;
    }
    subset.push_back(nums[current_index]);
    std::ranges::reverse(subset);
    return subset;
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.largestDivisibleSubset({1});
}
